src/Infrastructure.Styles.Analyzer/RMSYN1103ExtensionMethodsTestedAsNormalInvocationAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+﻿// This file is part of the re-motion Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// re-motion Framework is free software; you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation; either version 2.1 of the
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Infrastructure.Styles.Analyzer
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class RMSYN1103ExtensionMethodsTestedAsNormalInvocationSyntaxAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "RMSYN1103";
+
+        internal const string Title = "Extension methods are tested using normal method invocation syntax";
+        internal const string Message = "Extension methods should be tested with normal method invocation syntax";
+
+        internal const string Description =
+            "Extension methods should be tested using normal method invocation instead of by extension";
+
+        internal const string Category = "Readability";
+
+        private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            Title,
+            Message,
+            Category,
+            DiagnosticSeverity.Warning,
+            true,
+            Description);
+
+        private static readonly List<string> TestAttributeList = new List<string> { "Test", "TestAttribute" };
+
+        public override void Initialize (AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.EnableConcurrentExecution();
+
+            context.RegisterSyntaxNodeAction(AnalyzeInvocationExpression, SyntaxKind.InvocationExpression);
```

Comment:
I have performance concerns regarding the way the analyzer currently works. In a big solution there are a lot of invocation expressions and we execute the method analysis for all of them. Additionally we do the analysis for all projects regardless of them being a unit test project. I therefore suggest two improvements:

 - Register the syntax node action on a method do the "is test method" check and then search for all `InvocationExpression`s.
 - Only register the syntax node action if the project references NUnit

Reply:
Tried registering the Analyzation only when compilation analyzation output that it contains nunit, but it didnt continue and analyze all the delcarations, which kind of makes sense? Am not sure whether this is the way to do it :(

Reply:
Uhh yeah theres something else going wrong and i have no clue what it is.

Reply:
We should be able to use `RegisterCompilationStartAction` to hook into the start of the compilation, there we can check if `NUnit` is in the references and if it is we register the action.

Reply:
Sorry, forgot to push my changes, that was exactly what I did and exactly what didnt work.

Reply:
If you take a look at [our analyzer in re-motion](https://github.com/re-motion/Framework/blob/develop/Remotion/Web/Development.Analyzers/RMWEB0001_WebStringAnalyzer.cs) we do a similiar thing there. We require that certain types are available during compilation and only if they are there do we register the action. Then we also pass the resolved type in to be able to compare the types during the analyzer.

We should be able to do the same thing with the `TestAttribute` and use it to simplify the analyzer